File: tdc1/wk1/GroupAnagrams.java
package tdc1.wk1;

import java.util.*;

/**
 * Given an array of strings, group anagrams together.
 * <p>
 * Example:
 * <p>
 * Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
 * Output:
 * [
 * ["ate","eat","tea"],
 * ["nat","tan"],
 * ["bat"]
 * ]
 * <p>
 * Note:
 * <p>
 * All inputs will be in lowercase.
 * The order of your output does not matter.
 */
public class GroupAnagrams {
    public static List<List<String>> run() {
        String[] strings = {"eat", "tea", "tan", "ate", "nat", "bat"};
        List<List<String>> ans = groupAnagrams2(strings);
        for (List<String> list : ans) {
            System.out.print("[ ");
            for (String s : list) {
                System.out.print(s);
                System.out.print(' ');
            }
            System.out.print(" ]\n");
        }
        return ans;
    }

    /**
     * Build up map of sorted strings and their respective anagrams, then convert that to a
     * list of lists of strings.
     * <p>
     * O(nk*log n) time, O(nk) space
     * </p>
     * where n is the number of strings in str, and k is the maximum length of a string in strings
     * @param strings array of strings to be grouped
     * @return list of lists of grouped anagrams of strings in strings
     */
    public static List<List<String>> groupAnagrams1(String[] strings) {
        if (strings.length == 0) {
            return new ArrayList<>();
        }
        Map<String, List<String>> stringListMap = new HashMap<>();
        // build a map of sorted strings and strings
        // if str1 == str2 when they are both sorted
        for (String str : strings) {
            // yes, this is how you sort a string is Java: 3 lines and 2 type conversions...
            char[] arr = str.toCharArray();
            Arrays.sort(arr);
            String sortedStr = String.valueOf(arr);
            if (!stringListMap.containsKey(sortedStr)) {
                stringListMap.put(sortedStr, new ArrayList<>());
            }
            stringListMap.get(sortedStr).add(str);
        }
        // return the stringListMap as a list of lists of strings
        return new ArrayList<>(stringListMap.values());
    }

    /**
     * "hashing" approach, assuming 26-letter alphabet and '#' is a reserved character, this lets
     * us skip sorting each string.
     * <p>
     * O(nk) time, O(nk) space
     * </p>
     * where n is the number of strings in str, and k is the maximum length of a string in strings
     * @param strings array of strings to be grouped
     * @return list of lists of grouped anagrams of strings in strings
     */
    public static List<List<String>> groupAnagrams2(String[] strings) {
        if (strings.length == 0) return new ArrayList<>();
        Map<String, List<String>> ans = new HashMap<>();
        int[] count = new int[26];
        for (String s : strings) {
            Arrays.fill(count, 0);
            for (char c : s.toCharArray()) count[c - 'a']++;

            StringBuilder sb = new StringBuilder("");
            for (int i = 0; i < 26; i++) {
                sb.append('#');
                sb.append(count[i]);
            }
            String key = sb.toString();
            if (!ans.containsKey(key)) ans.put(key, new ArrayList<>());
            ans.get(key).add(s);
        }
        return new ArrayList<>(ans.values());
    }
}

File: tiq2/arraysandstrings/groupAnagrams.go
package arraysandstrings

import (
	"fmt"
	"sort"
)

// RunGroupAnagrams _
// Given an array of strings strs, group the anagrams together. You can return
// the answer in any order.
//
// An Anagram is a word or phrase formed by rearranging the letters of a
// different word or phrase, typically using all the original letters exactly
// once.
//
// Example 1:
//
// Input: strs = ["eat","tea","tan","ate","nat","bat"]
// Output: [["bat"],["nat","tan"],["ate","eat","tea"]]
//
// Example 2:
//
// Input: strs = [""]
// Output: [[""]]
//
// Example 3:
//
// Input: strs = ["a"]
// Output: [["a"]]
//
//
//
// Constraints:
//
//    1 <= strs.length <= 104
//    0 <= strs[i].length <= 100
//    strs[i] consists of lower-case English letters.
func RunGroupAnagrams() {
	// groupAnagrams2 should be faster than 1, but this cannot be clearly seen
	// in conventional tests
	group := groupAnagrams1
	tests := [][]string {
		{"eat","tea","tan","ate","nat","bat"},
		{""},
		{"a"},
		{"cab","tin","pew","duh","may","ill","buy","bar","max","doc"},
		{"dggg", "dgg"},
		{"dggg", "dgg"},
	}
	for _, strs := range tests {
		g := group(strs)
		fmt.Println(g)
	}
}

// O(n mlogm) time where m is the length of the longest string in the array
// O(nm) space since the map stores up to n things, each thing
// (only the key matters since its the extra bit, values are references) is m
// or shorter
func groupAnagrams1(strs []string) [][]string {
	anagrams := make(map[string][]string)
	for _, str := range strs {
		r := []rune(str)
		sort.Slice(r, func(i, j int) bool {
			return r[i] < r[j]
		})
		sorted := string(r)
		anagrams[sorted] = append(anagrams[sorted], str)
	}
	ans := make([][]string, 0)
	for _, a := range anagrams {
		ans = append(ans, a)
	}
	return ans
}

// charCount counts the number of occurences of lower-case English letters in
// str. It is given that strs[i] consists of lower-case English letters.
// O(n) time
// O(1) (extra) space
func charCounts(str string) [26]int{
	hashable := [26]int{}
	for _, r := range str {
		pos := r - 'a'
		hashable[pos]++
	}
	return hashable
}

// O(nm) time where m is the length of the longest string
// O(nm) space
//
// Instead of sorting, make use of lower-case English letter constraint and
// use character count arrays
func groupAnagrams2(strs []string) [][]string {
	anagrams := make(map[[26]int][]string)
	for _, str := range strs {
		h := charCounts(str)
		anagrams[h] = append(anagrams[h], str)
	}
	ans := make([][]string, 0)
	for _, a := range anagrams {
		ans = append(ans, a)
	}
	return ans
}
